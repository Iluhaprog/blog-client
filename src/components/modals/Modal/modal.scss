@import '../../../vars.scss';

@keyframes close {
    0% {
        opacity: 1;
    }
    90% {
        opacity: 0;
    }
    100% {
        display: none;
    }
}
@keyframes view {
    0% {
        display: block;
        opacity: 0;
    }
    10% {
        opacity: 0;
    }
    100% {
        opacity: 1;
    }
}

.modal {
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    background-color: $text-color-4;
    display: none;
    transition: .3s ease-in-out;

    .row {
        height: 100%;
    }

    &_visible {
        display: block;
    }

    &-window {
        max-width: 300px;
        min-height: 100px;
        width: 100%;
        background-color: $text-color-1;
        border-radius: 5px;
        box-shadow: $shadow;
        padding: 15px 10px;

        &__text {
            font-size: 14px;
            margin: 0;
            color: $text-color-8;
            margin-bottom: 15px;
        }

        &__input {
            width: 100%;
            font-size: 1em;
            padding: 0.5em;
            margin-bottom: 15px;
            border-radius: 5px;
            border: $border-2;
            transition: .3s ease-in-out;
            outline: none;

            &:focus {
                border: $border-1;
            }
        }

        strong {
            display: block;
            color: $error;
            margin-bottom: 10px;
        }
    }
}