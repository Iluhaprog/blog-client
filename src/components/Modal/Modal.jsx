import React from 'react';
import {clearModal} from '../../store/modal/modalActions';
import {connect} from 'react-redux';
import {ModalWrapper} from './Styled/ModalWrapper';
import PropTypes from 'prop-types';
import {ModalScreenTypes} from '../../store/modal/ModalFormTypes';
import {DefaultScreen} from './Screens/Default';
import {FileSelectScreen} from './Screens/FileSelect';
import {InputScreen} from './Screens/Input';

const ModalScreens = {
  [ModalScreenTypes.NONE]: DefaultScreen,
  [ModalScreenTypes.FILE_SELECT]: FileSelectScreen,
  [ModalScreenTypes.INPUT]: InputScreen,
};

let Modal = ({modal, theme, close}) => {
  const Screen = ModalScreens[modal.formType];

  return (
    <ModalWrapper isVisible={modal.isVisible}>
      <Screen
        modal={modal}
        theme={theme}
        onClose={close}
        onSuccess={modal.successHandler}
      />
    </ModalWrapper>
  );
};

Modal.propTypes = {
  modal: PropTypes.object,
  theme: PropTypes.string,
  close: PropTypes.func,
};

const mapStateToProps = (state) => ({
  modal: state.modal,
  theme: state.settings.theme,
});

const mapDispatchToProps = (dispatch) => ({
  close: () => {
    dispatch(clearModal());
  },
});

Modal = connect(mapStateToProps, mapDispatchToProps)(Modal);

export {Modal};
