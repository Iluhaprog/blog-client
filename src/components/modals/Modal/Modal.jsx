import React, { useState } from 'react';
import { DangerButton, SuccessButton } from '../../buttons';
import { Row } from '../../containers';
import './modal.scss';

function Modal(props) {
    const { modal, close, setFail} = props;
    const { text, handleSuccess, fail, withInput, visible } = modal;
    const [value, setValue] = useState('');

    return (
        <div className={`modal ${visible ? 'modal_visible' : ''}`}>
            <Row justifyContent='c' alignItems='c'>
                <div className='modal-window'>
                    <p className='modal-window__text'>
                        {text}
                    </p>
                    {
                        withInput 
                            ? (
                                <input 
                                    type='text' 
                                    className='modal-window__input'
                                    onChange={e => setValue(e.target.value)}
                                    value={value}
                                />
                              )
                            : ''
                    }
                    {
                        fail.message 
                            ? (<strong>{fail.message}</strong>)
                            : '' 
                    }
                    <Row justifyContent='sb'>
                        <SuccessButton 
                            text='ok'
                            onClick={() => {
                                handleSuccess(value, setFail);
                                close();
                            }}
                        />
                        <DangerButton 
                            text='close' 
                            onClick={() => {
                                close();
                                setValue('');
                            }}
                        />
                    </Row>
                </div>
            </Row>
        </div>
    );
}

export default Modal;