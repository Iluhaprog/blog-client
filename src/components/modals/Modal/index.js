import { connect } from 'react-redux';
import { clearModal, setFail } from '../../../actoins/modal';
import m from './Modal';

const mapStateToProps = state => ({
    modal: state.modal,
});

const mapDispatchToProps = dispatch => ({
    close: () => dispatch(clearModal()),
    setFail: message => dispatch(setFail(message)),
});

const Modal = connect(mapStateToProps, mapDispatchToProps)(m);

export {
    Modal,
};