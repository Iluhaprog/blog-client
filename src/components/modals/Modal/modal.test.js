import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import Modal from './Modal';

const getProps = (props = {}) => ({
    modal: {
        text: props.text || '',
        withInput: props.withInput || false,
        handleSuccess: props.handleSuccess || (() => {}),
        visible: props.visible || false,
        fail: props.fail || { message: '' },
    },
    close: props.close || (() => {}),
    setFail: () => {},
});

describe('Test modal view', () => {
    test('Should return unvisible modal', () => {
        const props = getProps();
        const { asFragment } = render(<Modal {...props} />);
        expect(asFragment()).toMatchSnapshot();
    });
    test('Should return visible modal', () => {
        const props = getProps({ visible: true });
        const { asFragment } = render(<Modal {...props} />);
        expect(asFragment()).toMatchSnapshot();
    });
    test('Should return modal with input', () => {
        const props = getProps({ withInput: true });
        const { asFragment } = render(<Modal {...props} />);
        expect(asFragment()).toMatchSnapshot();
    });
    test('Should return modal with fail', () => {
        const props = getProps({ fail: { message: 'test fail' } });
        const { asFragment } = render(<Modal {...props} />);
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('Test modal handlers', () => {
    test('Should call handleSuccess function', async () => {
        const handleSuccess = jest.fn();
        const props = getProps({ handleSuccess });
        const { container } = render(<Modal {...props} />);
        fireEvent.click(container.querySelector('.success-button'));
        await waitFor(() => {
            expect(handleSuccess).toHaveBeenCalledTimes(1);
        });
    });
    test('Should call close function', async () => {
        const close = jest.fn();
        const props = getProps({ close });
        const { container } = render(<Modal {...props} />);
        fireEvent.click(container.querySelector('.danger-button'));
        await waitFor(() => {
            expect(close).toHaveBeenCalledTimes(1);
        });
    });
});