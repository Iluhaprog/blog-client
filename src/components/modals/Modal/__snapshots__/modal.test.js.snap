// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test modal view Should return modal with fail 1`] = `
<DocumentFragment>
  <div
    class="modal "
  >
    <div
      class="row row_jc-c row_ai-c "
    >
      <div
        class="modal-window"
      >
        <p
          class="modal-window__text"
        />
        <strong>
          test fail
        </strong>
        <div
          class="row row_jc-sb  "
        >
          <button
            class="button default-button success-button"
          >
            ok
          </button>
          <button
            class="button default-button danger-button"
          >
            close
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Test modal view Should return modal with input 1`] = `
<DocumentFragment>
  <div
    class="modal "
  >
    <div
      class="row row_jc-c row_ai-c "
    >
      <div
        class="modal-window"
      >
        <p
          class="modal-window__text"
        />
        <input
          class="modal-window__input"
          type="text"
          value=""
        />
        <div
          class="row row_jc-sb  "
        >
          <button
            class="button default-button success-button"
          >
            ok
          </button>
          <button
            class="button default-button danger-button"
          >
            close
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Test modal view Should return unvisible modal 1`] = `
<DocumentFragment>
  <div
    class="modal "
  >
    <div
      class="row row_jc-c row_ai-c "
    >
      <div
        class="modal-window"
      >
        <p
          class="modal-window__text"
        />
        <div
          class="row row_jc-sb  "
        >
          <button
            class="button default-button success-button"
          >
            ok
          </button>
          <button
            class="button default-button danger-button"
          >
            close
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Test modal view Should return visible modal 1`] = `
<DocumentFragment>
  <div
    class="modal modal_visible"
  >
    <div
      class="row row_jc-c row_ai-c "
    >
      <div
        class="modal-window"
      >
        <p
          class="modal-window__text"
        />
        <div
          class="row row_jc-sb  "
        >
          <button
            class="button default-button success-button"
          >
            ok
          </button>
          <button
            class="button default-button danger-button"
          >
            close
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
